File: gateway-server/src/main/java/org/lucycato/gatewayserver/application/service/LoggingGlobalFilterService.java
package org.lucycato.gatewayserver.application.service;

import lombok.extern.slf4j.Slf4j;
import org.lucycato.gatewayserver.application.port.in.LoggingGlobalFilterUseCase;
import org.lucycato.gatewayserver.application.port.in.command.LogRequestCommand;
import org.lucycato.gatewayserver.application.port.in.command.LogResponseCommand;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Slf4j
@Service
public class LoggingGlobalFilterService implements LoggingGlobalFilterUseCase {

    @Override
    public Mono<Void> logRequest(LogRequestCommand command) {
        return Mono.fromRunnable(() -> {
            String originalUri = (command.getUris().isEmpty()) ? "Unknown" : command.getUris().iterator().next().toString();
            String routeUri = "lb://" + command.getRouteServiceId() + command.getRouteUriPath();

            StringBuilder headers = new StringBuilder();
            command.getHeaders().forEach((headerName, headerValues) -> {
                headers
                        .append("   - ")
                        .append(headerName)
                        .append(": ")
                        .append(String.join(", ", headerValues))
                        .append("\n");
            });

            log.info("\n>> -- >> -- >> -- >> REQUEST  >> -- >> -- >> -- >>\n"
                            + "Original URL: {}\n"
                            + "Route URL: {}\n"
                            + "Jwt: {}\n"
                            + "Method: {}\n"
                            + "Headers:\n{}\n"
                            + ">> -- >> -- >> -- >> REQUEST  >> -- >> -- >> -- >>\n",
                    originalUri,
                    routeUri,
                    command.getToken(),
                    command.getMethod(),
                    headers
            );
        });
    }

    @Override
    public Mono<Void> logResponse(LogResponseCommand command) {
        return Mono.fromRunnable(() -> {
            StringBuilder headers = new StringBuilder();
            command.getHeaders().forEach((headerName, headerValues) -> {
                headers
                        .append("   - ")
                        .append(headerName)
                        .append(": ")
                        .append(String.join(", ", headerValues))
                        .append("\n");
            });

            log.info("\n<< -- << -- << -- << RESPONSE << -- << -- << -- <<\n"
                            + "Status: {}\n"
                            + "Headers:\n{}\n"
                            + "<< -- << -- << -- << RESPONSE << -- << -- << -- <<\n",
                    command.getStatusCode(),
                    headers
            );
        });
    }
}

File: gateway-server/src/main/kotlin/org/lucycato/gatewayservice/application/service/LoggingGlobalFilterService.kt
package org.lucycato.gatewayservice.application.service

import org.lucycato.gatewayservice.application.port.`in`.LoggingGlobalFilterUseCase
import org.lucycato.gatewayservice.application.port.`in`.commnad.LogRequestCommand
import org.lucycato.gatewayservice.application.port.`in`.commnad.LogResponseCommand
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import reactor.core.publisher.Mono

@Service
class LoggingGlobalFilterService : LoggingGlobalFilterUseCase {
    private val log: Logger = LoggerFactory.getLogger(this.javaClass)
    override fun logRequest(command: LogRequestCommand): Mono<Void> {
        return Mono.fromRunnable {
            val originalUri = if (command.uris.isEmpty()) "Unknown" else command.uris.iterator().next().toString()
            val routeUri = "lb://${command.routeServiceId}${command.routeUriPath}"

            val headers = StringBuilder().apply {
                command.headers.forEach { (headerName, headerValues) ->
                    append("   - ")
                    append(headerName)
                    append(": ")
                    append(headerValues.joinToString(", "))
                    append("\n")
                }
            }

            log.info("""
            
            >> -- >> -- >> -- >> REQUEST  >> -- >> -- >> -- >>
            Original URL: $originalUri
            Route URL: $routeUri
            Jwt: ${command.token}
            Method: ${command.method}
            Headers:
            $headers
            >> -- >> -- >> -- >> REQUEST  >> -- >> -- >> -- >>
            
            """.trimIndent()
            )
        }
    }

    override fun logResponse(command: LogResponseCommand): Mono<Void> {
        return Mono.fromRunnable {
            val headers = StringBuilder().apply {
                command.headers.forEach { (headerName, headerValues) ->
                    append("   - ")
                    append(headerName)
                    append(": ")
                    append(headerValues.joinToString(", "))
                    append("\n")
                }
            }

            log.info("""
                
            << -- << -- << -- << RESPONSE << -- << -- << -- <<
            Status: ${command.statusCode}
            Headers:
            $headers
            << -- << -- << -- << RESPONSE << -- << -- << -- <<
            
            """.trimIndent()
            )
        }
    }
}